rust/src/adapter/datadog.rs

```diff
@@ -0,0 +1,160 @@
+};
+use serde_json::json;
+use tokio::sync::Mutex;
+use ureq;
+use url::Url;
+
+use super::{
+    datadog_formatter::{self, Trace, Span},
+    Adapter, Collector,
+};
+
+pub type DatadogAdapterContainer = Arc<Mutex<DatadogAdapter>>;
+
+#[derive(Clone)]
+pub struct DatadogAdapter {
+    pub trace_id: u64,
+    pub spans: Vec<Span>,
+    pub config: DatadogConfig,
+}
+
+#[derive(Clone)]
+pub struct DatadogConfig {
+    pub agent_host: String,
+    pub service_name: String,
+    pub default_tags: HashMap<String, String>,
+    pub trace_type: String,
+}
+
+impl DatadogConfig {
+    pub fn new() -> DatadogConfig {
+        DatadogConfig {
+            agent_host: "http://localhost:8126".into(),
+            service_name: "my-wasm-service".into(),
+            default_tags: HashMap::new(),
+            trace_type: "web".into(),
+        }
+    }
+}
+
+impl DatadogAdapter {
+    pub fn new(config: DatadogConfig) -> DatadogAdapterContainer {
+        let adapter = DatadogAdapter {
+            trace_id: datadog_formatter::new_span_id(),
+            spans: vec![],
+            config,
+        };
+
+        Arc::new(Mutex::new(adapter))
+    }
+
+    pub fn new_collector(&mut self) -> usize {
+        next_id()
+    }
+
+    fn _handle_event(&mut self, event: Event, parent_id: Option<u64>) -> Option<Vec<Span>> {
+        match event {
+            Event::Func(_id, f) => {
+                let function_name = &f.name.clone().unwrap_or("unknown-name".to_string());
+                let name = format!("function-call-{}", &function_name);
+
+                let config = self.config.clone();
+                let span =
+                    Span::new(config, self.trace_id.clone(), parent_id, name, f.start, f.end);
+
+                let span_id = Some(span.span_id.clone());
+                let mut spans = vec![span];
+
+                for e in f.within.iter() {
+                    if let Some(mut s) = self._handle_event(e.to_owned(), span_id.clone()) {
+                        spans.append(&mut s);
+                    };
+                }
+
+                Some(spans)
+            }
+            Event::Alloc(_id, a) => {
+                let config = self.config.clone();
+                let span = Span::new(
+                    config,
+                    self.trace_id,
+                    parent_id,
+                    "allocation".to_string(),
+                    a.ts,
+                    a.ts,
+                );
+                Some(vec![span])
+            }
+            Event::Metadata(_id, Metadata { key, value }) => {
+                if key == "trace_id" {
+                    self.trace_id = value;
+                }
+                None
+            }
+            Event::Shutdown(_id) => {
+                // when we receive the shutdown
+                // then dump the trace to the agent
+                self.shutdown();
+                self.spans.clear();
+                None
+            },
+        }
+    }
```

Comment:
If this is pretty standard across these adapters, we can add this as a [Default impl](https://doc.rust-lang.org/book/ch10-02-traits.html#default-implementations) in the trait definition itself.